src/main/java/fr/xephi/authme/listener/PlayerListener.java

```diff
@@ -190,46 +192,73 @@ public void onPlayerJoin(PlayerJoinEvent event) {
+        final boolean isAuthAvailable = auth != null;
+        onJoinVerifier.checkKickNonRegistered(isAuthAvailable);
+        onJoinVerifier.checkAntibot(name, isAuthAvailable);
+        onJoinVerifier.checkNameCasing(name, auth);
+        onJoinVerifier.checkPlayerCountry(name, ip, isAuthAvailable);
+    }
+
     // Note #831: AsyncPlayerPreLoginEvent is not fired by all servers in offline mode
     // e.g. CraftBukkit does not fire it. So we need to run crucial things with PlayerLoginEvent.
     // Single session feature can be implemented for Spigot and CraftBukkit by canceling a kick
     // event caused by "logged in from another location". The nicer way, but only for Spigot, would be
     // to check in the AsyncPlayerPreLoginEvent. To support all servers, we use the less nice way.
 
+    private boolean isAsyncPlayerPreLoginEventCalled = false;
+
+    @EventHandler(priority = EventPriority.HIGHEST)
+    public void onAsyncPlayerPreLoginEvent(AsyncPlayerPreLoginEvent event) {
+        isAsyncPlayerPreLoginEventCalled = true;
+
+        final String name = event.getName();
+
+        if (validationService.isUnrestricted(name)) {
+            return;
+        }
+
+        try {
+            runOnJoinChecks(name, event.getAddress().getHostAddress());
+        } catch (FailedVerificationException e) {
+            event.setKickMessage(m.retrieveSingle(e.getReason(), e.getArgs()));
+            event.setLoginResult(AsyncPlayerPreLoginEvent.Result.KICK_OTHER);
+        }
+    }
+
     @EventHandler(priority = EventPriority.LOW)
     public void onPlayerLogin(PlayerLoginEvent event) {
         final Player player = event.getPlayer();
         final String name = player.getName();
+
         if (validationService.isUnrestricted(name)) {
             return;
         }
+
         if (onJoinVerifier.refusePlayerForFullServer(event)) {
             return;
         }
         if (event.getResult() != PlayerLoginEvent.Result.ALLOWED) {
             return;
         }
 
-        try {
-            // Fast stuff
-            onJoinVerifier.checkSingleSession(name);
-            onJoinVerifier.checkIsValidName(name);
-
-            // Get the auth later as this may cause the single session check to fail
-            // Slow stuff
-            final PlayerAuth auth = dataSource.getAuth(name);
-            final boolean isAuthAvailable = auth != null;
-            onJoinVerifier.checkKickNonRegistered(isAuthAvailable);
-            onJoinVerifier.checkAntibot(player, isAuthAvailable);
-            onJoinVerifier.checkNameCasing(player, auth);
-            onJoinVerifier.checkPlayerCountry(player, event.getAddress().getHostAddress(), isAuthAvailable);
-        } catch (FailedVerificationException e) {
-            event.setKickMessage(m.retrieveSingle(e.getReason(), e.getArgs()));
-            event.setResult(PlayerLoginEvent.Result.KICK_OTHER);
-            return;
-        }
-
         teleportationService.teleportOnJoin(player);
+
```

Comment:
This introduces a change in behavior I'm not sure is intentional—previously with the red block we did all of the onJoin verifications before and kicked the player. With the green addition the player gets teleported before any checks are made.
This is maybe better because this way a player is teleported as fast as possible (since we don't want to show the original location before login in many cases) ? But at the same time in the case of a bot attack we might be teleporting a lot of players before determining that they have no right being online. Difficult...

Reply:
Teleporting players to a same spot shouldn't be resource intensive

Reply:
Agreed ✔️